### PR TITLE
Remove constraints for private interface method implementations

### PR DESCRIPTION
**Pull request by mattwar**
Fix for #3928

C# explicit interface implementations cannot specify constraints. This changes makes it so any method declaration being converted to a explicit implementation via AsPrivateInterfaceImplementation has any/all constraints removed.

@Pilchie @dpoeschl @rchande @DustinCampbell @CyrusNajmabadi please review

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 1]
```diff
@@ -367,11 +368,24 @@ public override SyntaxNode AsPublicInterfaceImplementation(SyntaxNode declaratio
...
+            void Moo<T>() where T : IEnumerable<T>;
+        }
+
+        public class Bar : Foo
+        {
+            public void Moo<T>() where T : IEnumerable<T>
+            {
+            }
+        }
+
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Was this you doing some testing?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -357,6 +357,7 @@ public override SyntaxNode AsPublicInterfaceImplementation(SyntaxNode declaratio
             return PreserveTrivia(declaration, d =>
             {
                 d = AsImplementation(d, Accessibility.Public);
+                d = WithoutConstraints(d);
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: This is in 'AsPublicInterfaceImplementation`.  Do you want to remove constraints in this case?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (mattwar) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Experimenting.  But don't you think these types need to live on in the tests?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 2 above)
Comment (mattwar) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Nope..

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Apart from the name choice, they could live on in the tests, but we probably shouldn't have them here in the product code :smile:

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 3]
```diff
@@ -382,6 +396,20 @@ public override SyntaxNode AsPrivateInterfaceImplementation(SyntaxNode declarati
             });
         }
 
+        private SyntaxNode WithoutConstraints(SyntaxNode declaration)
+        {
+            if (declaration.IsKind(SyntaxKind.MethodDeclaration))
+            { 
+                var method = (MethodDeclarationSyntax)declaration;
+                if (method.ConstraintClauses.Count > 0)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: This check seems unnecessary.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 4]
```diff
@@ -382,6 +396,20 @@ public override SyntaxNode AsPrivateInterfaceImplementation(SyntaxNode declarati
...
+                {
+                    return this.RemoveNodes(method, method.ConstraintClauses);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: i.e. 'RemoveNodes' should check if the node list is empty, and just immediately bail if it is.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 5]
```diff
@@ -367,11 +368,24 @@ public override SyntaxNode AsPublicInterfaceImplementation(SyntaxNode declaratio
...
+        }
+
+        public class Bar : Foo
+        {
+            public void Moo<T>() where T : IEnumerable<T>
+            {
+            }
+        }
+
         public override SyntaxNode AsPrivateInterfaceImplementation(SyntaxNode declaration, SyntaxNode interfaceTypeName, string interfaceMemberName)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Why the term 'Private' here?  I thought this was generally called an 'ExplicitInterfaceImplementation'.  That's also how we present this concept through features (i.e. "implement this interface 'explicitly'").

Comment (mattwar) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: That's true for C#, but not for VB.  VB always implements interface methods explicitly.

**Comment by dpoeschl** [comment]
:+1:

**Pull request merged**